feat: Handle missing original bookingReference or invalid reference during reschedule - Create booking anyway

Devanshusharma2005 opened the pull request:
## What does this PR do?

This PR fixes the issue where rescheduling a booking fails if the original Google Calendar event is missing or has an empty UID. The solution implements a **declarative approach**: after rescheduling, a valid calendar event will always exist, regardless of whether one existed before.

- Fixes #22192
- Fixes CAL-6020

### Key Changes

- **EventManager.reschedule()**: Updated to create new calendar events when none exist (missing references or empty UIDs)
- **New service classes**: Extracted `BookingReferenceService` and `EventResultService` for cleaner separation of concerns
- **Comprehensive test coverage**: Added tests for missing calendar event scenarios and empty UID edge cases  
- **Booking reference management**: Properly updates booking references with newly created calendar events while preserving existing non-calendar references

### Technical Implementation

The fix follows this declarative approach:
1. **If valid calendar event exists** (has a non-empty UID) → update it as before
2. **If no valid calendar event exists** (missing or empty UID) → create new calendar events using `createAllCalendarEvents()`
3. **Update booking references** with new calendar event information
4. **Preserve existing references** like video meeting references

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    handleNewBooking["handleNewBooking.ts<br/>(reschedule flow)"]:::context
    EventManager["EventManager.ts<br/>reschedule()"]:::major-edit
    rescheduleTest["reschedule.test.ts<br/>(new test cases)"]:::major-edit
    
    handleNewBooking --> EventManager
    EventManager --> createAllCalendarEvents["createAllCalendarEvents()"]:::context
    EventManager --> updateAllCalendarEvents["updateAllCalendarEvents()"]:::context
    
    rescheduleTest -.-> EventManager
    
    EventManager --> BookingReferences["Booking References<br/>(calendar + non-calendar)"]:::minor-edit
    
    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit  
        L3[Context/No Edit]:::context
    end
    
    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB
    classDef context fill:#FFFFFF
```

### Updates since last revision

- Applied optional chaining fix for `bookingCalendarReference?.uid` check (CodeRabbit review feedback)
- Merged latest main branch to resolve conflicts

## Visual Demo (For contributors especially)

**Note**: A demo/recording showing before and after behavior was requested by @hariombalhara. This would help validate the fix works correctly with real calendar integrations.

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. N/A - no documentation changes needed.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

**Recommended Test Plan**: 
1. Create a booking that fails calendar event creation initially (simulate by removing calendar credentials temporarily)
2. Restore calendar credentials and attempt to reschedule the booking
3. Verify a valid calendar event now exists and future reschedules work normally

**Automated tests added**:
- Test for rescheduling booking with no existing calendar event UID
- Test for rescheduling booking with empty calendar event UID

**Manual testing checklist**:
- [ ] End-to-end testing with real Google Calendar integration
- [ ] Verify existing bookings with valid calendar events still reschedule correctly
- [ ] Test with multiple calendar integrations

## Checklist

- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have checked if my changes generate no new warnings

---


**Link to Devin run**: https://app.devin.ai/sessions/abf04a80181e448e887ecb8480d3d6b7
**Original session**: https://app.devin.ai/sessions/4127248fd8474da3852f8b4c4ace1d18  
**Requested by**: @Devanshusharma2005 (original author), completed by Devin
<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/calcom/cal.com/pull/22660" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

devin-ai-integration[bot] committed:
feat: implement declarative calendar event handling in reschedule

- Fix EventManager.reschedule to create new calendar events when none exist
- Add comprehensive tests for missing and empty calendar event UID scenarios
- Ensure valid calendar event always exists after rescheduling
- Update booking references with newly created calendar event information

Fixes the issue where rescheduling fails if original Google Calendar event is missing

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

devin-ai-integration[bot] committed:
fix: add missing calendar service mocking to resolve CI test failures

- Add mockCalendarToHaveNoBusySlots calls with proper create parameters to tests that use Google Calendar
- Fixes 'Unimplemented appStoreMock' errors in CI
- Ensures tests that trigger createAllCalendarEvents have proper mocking

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

Devanshusharma2005 committed:
Merge branch 'main' into devin/calendar-reschedule-declarative-1753086185

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into devin/calendar-reschedule-declarative-1753086185

hariombalhara commented on packages/lib/EventManager.ts:
```diff
@@ -643,10 +648,34 @@ export default class EventManager {
+            // Update existing calendar events
             results.push(...(await this.updateAllCalendarEvents(evt, booking, newBookingId)));
+          } else {
+            log.debug("No valid calendar event found for booking, creating new calendar events");
+            const createdEvents = await this.createAllCalendarEvents(evt);
+            results.push(...createdEvents);
+
+            // Update booking references with newly created calendar events
+            const calendarReferences = createdEvents
+              .filter((result) => result.type.includes("_calendar") && result.success)
+              .map((result) => ({
+                type: result.type,
+                uid: result.uid,
+                meetingId: result.uid,
+                meetingPassword: result.originalEvent.attendees?.[0]?.timeZone || "",
+                meetingUrl: result.originalEvent.videoCallData?.url || "",
+                externalCalendarId: result.externalId || null,
+                credentialId: result.credentialId || null,
+              }));
```
The _reschedule_ fn is already doing a lot of things violating Single Responsibility principle. Let's do this handling in createMissingCalendarEvents or handleMissingCalendarEvents

hariombalhara commented on packages/lib/EventManager.ts:
```diff
@@ -596,7 +597,11 @@ export default class EventManager {
     }
 
     const shouldUpdateBookingReferences =
```
This variable seems to be unused now.

Devanshusharma2005 commented on packages/lib/EventManager.ts:
```diff
@@ -643,10 +648,34 @@ export default class EventManager {
+            const calendarReferences = createdEvents
+              .filter((result) => result.type.includes("_calendar") && result.success)
+              .map((result) => ({
+                type: result.type,
+                uid: result.uid,
+                meetingId: result.uid,
+                meetingPassword: result.originalEvent.attendees?.[0]?.timeZone || "",
+                meetingUrl: result.originalEvent.videoCallData?.url || "",
+                externalCalendarId: result.externalId || null,
+                credentialId: result.credentialId || null,
```
Sounds good. Dropping commit to the above suggestions now.

hariombalhara committed:
refactor: extract BookingReferenceService.buildFromResult() from EventManager

hariombalhara committed:
refactor: rename bookingReference.ts to BookingReferenceService.ts

hariombalhara reviewed (changes requested):
Some comments need to be addressed

devin-ai-integration[bot] committed:
refactor: extract handleMissingCalendarEvents method from EventManager.reschedule

- Extract lines 656-671 into new handleMissingCalendarEvents method
- Improve code organization following Single Responsibility Principle
- Preserve original behavior and typing
- Add proper TypeScript typing for filter function parameter
- Clean up variable usage and improve maintainability

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

devin-ai-integration[bot] committed:
resolve merge conflict: preserve handleMissingCalendarEvents refactoring with BookingReferenceService

- Keep refactored handleMissingCalendarEvents method structure
- Adopt BookingReferenceService.buildFromResult usage from remote changes
- Maintain Single Responsibility Principle improvements
- Preserve all existing functionality and typing

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

hariombalhara committed:
Use services

Devanshusharma2005 committed:
Merge branch 'main' into devin/calendar-reschedule-declarative-1753086185

Devanshusharma2005 committed:
finalshouldUpdateBookingReferences -> shouldUpdateBookingReferences

Devanshusharma2005 committed:
Merge branch 'main' into devin/calendar-reschedule-declarative-1753086185

anikdhabal committed:
Merge branch 'main' into devin/calendar-reschedule-declarative-1753086185

hariombalhara reviewed (changes requested):
We are missing a demo(before and after) here. I will be helping @Devanshusharma2005 with that.

devin-ai-integration[bot] committed:
Merge branch 'main' into devin/calendar-reschedule-declarative-1753086185

Resolved merge conflicts in:
- packages/features/bookings/lib/EventManager.ts
- packages/features/bookings/lib/handleNewBooking/test/reschedule.test.ts

Conflict resolution:
- EventManager.ts: Merged new BookingReferenceService and EventResultService imports with updated prisma import style, kept new class properties with Record<string, any> for appOptions, and used main's better structured nested if statements for reschedule logic
- reschedule.test.ts: Kept new expectSuccessfulCalendarEventCreationInCalendar import with updated @calcom/testing import paths

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
refactor: use optional chaining for bookingCalendarReference check

Address CodeRabbit review comment to use optional chaining instead of
explicit null check for better readability and safety.

Co-authored-by: Devanshusharma2005 <devanshusharma2005@users.noreply.github.com>
Co-Authored-By: unknown <>

sahitya-chandra committed:
Merge branch 'main' into devin/calendar-reschedule-declarative-1753086185

sahitya-chandra committed:
Merge branch 'main' into devin/calendar-reschedule-declarative-1753086185

sahitya-chandra committed:
refactor: remove BookingReferenceService and EventResultService; update EventManager to handle calendar events directly

- Deleted BookingReferenceService and EventResultService to streamline event management.
- Updated EventManager to directly manage calendar event references and results.
- Enhanced handling of new calendar events with a new flag for created events.
- Simplified test cases by removing unnecessary mock data for calendar slots.
